### DISCUSSION TITLE
RuntimeError: Form data requires "python-multipart" to be installed.

### DISCUSSION THREAD

**Discussion by Honzu-Code** [fastapi/fastapi#5144] [Questions] [answered]
Hi guys i have problem with Form multipart. I tested in visual studio code and in normal terminal on linux ubuntu 21.04.
idk what to do i tryed install (pip install multipart) and (pip install python-multipart) but fastapi doesn't see this multipart i think from virtual environment. 
![image](https://user-images.githubusercontent.com/66671802/178961159-0abd33b5-408b-4077-8f0b-978f1029ab8d.png)
and pip list:
![image](https://user-images.githubusercontent.com/66671802/178961343-9468c86f-a006-4e15-b7f5-cb4cf35bdec6.png)
i tested on python 3.10 and 3.9.5

**Comment by Honzu-Code** [comment]
When iam running from global terminal this python package work :) can you guys fix this for vs code?

**Reply by iudeen** [reply]
Are you running your FastAPI app from same environment as you are installing? I don't think so. 

Also, FastAPI has nothing to do with VSCode or anything, its rather your Python Interpreter. 

Check the Python env.

**Reply by Honzu-Code** [reply]
I run this on global terminal and work. Visual Studio Code have bug with this i think

**Reply by iudeen** [reply]
There is no bug. We do use virtual environment within Visual Studio code. And never faced such an issue.

Just make sure the vs code terminal has the appropriate venv activated.

**Reply by kw-fp** [reply]
@Honzu-Code I had the same issue. Try `pip install wheel` in your virtual environment before `pip install python-multipart`. Should work.

**Reply by Honzu-Code** [reply]
I will try

**Reply by lifsys** [reply]
This happened to me as well.  You likely had uvicorn installed on base environment.  When you created your new environment you installed fastapi, but not uvicorn.  Looking at your error and path information, it was seeking fastapi/uvicorn in base.  This is why it works in "global terminal".  Install fastapi AND uvicorn in your env.

**Comment by tiangolo** [comment, marked as answer]
If you install both you will have problems because they conflict, they expose the same name. Uninstall `multipart` and only keep `python-multipart`.

**Reply by ghost** [reply]
Great! Works for me now. Thanks 🙏

**Comment by Alan3344** [comment]
My envirement：
```bash
(venv) ~/item/api-server python          
Python 3.11.5 (main, Aug 24 2023, 15:09:45) [Clang 14.0.3 (clang-1403.0.22.14.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> exit()
(venv) ~/item/api-server pip -V          
pip 23.3.1 from /Users/alan/item/api-server/venv/lib/python3.11/site-packages/pip (python 3.11)
(venv) ~/item/api-server whereis uvicorn 
uvicorn: /Users/alan/item/api-server/venv/bin/uvicorn
(venv) ~/item/api-server 

```

For me, I confirmed that the virtual environment was turned on
And there is no such package in the environment I confirmed. `multipart`

![image](https://github.com/tiangolo/fastapi/assets/60773912/f640f3b3-66d8-4717-86c6-b78412a3346b)

But:
<details>

(venv) ~/item/api-server uvicorn main:app --reload
INFO:     Will watch for changes in these directories: ['/Users/alan/item/api-server']
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [96608] using WatchFiles
Form data requires "python-multipart" to be installed. 
You can install "python-multipart" with: 

pip install python-multipart

Process SpawnProcess-1:
Traceback (most recent call last):
  File "/opt/homebrew/Cellar/python@3.11/3.11.5/Frameworks/Python.framework/Versions/3.11/lib/python3.11/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/opt/homebrew/Cellar/python@3.11/3.11.5/Frameworks/Python.framework/Versions/3.11/lib/python3.11/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/homebrew/lib/python3.11/site-packages/uvicorn/_subprocess.py", line 76, in subprocess_started
    target(sockets=sockets)
  File "/opt/homebrew/lib/python3.11/site-packages/uvicorn/server.py", line 61, in run
    return asyncio.run(self.serve(sockets=sockets))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.5/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.5/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "uvloop/loop.pyx", line 1517, in uvloop.loop.Loop.run_until_complete
  File "/opt/homebrew/lib/python3.11/site-packages/uvicorn/server.py", line 68, in serve
    config.load()
  File "/opt/homebrew/lib/python3.11/site-packages/uvicorn/config.py", line 467, in load
    self.loaded_app = import_from_string(self.app)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/uvicorn/importer.py", line 21, in import_from_string
    module = importlib.import_module(module_str)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.5/Frameworks/Python.framework/Versions/3.11/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/Users/alan/item/api-server/main.py", line 11, in <module>
    @file.post('/upload')
     ^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/fastapi/routing.py", line 704, in decorator
    self.add_api_route(
  File "/opt/homebrew/lib/python3.11/site-packages/fastapi/routing.py", line 643, in add_api_route
    route = route_class(
            ^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/fastapi/routing.py", line 495, in __init__
    self.body_field = get_body_field(dependant=self.dependant, name=self.unique_id)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/fastapi/dependencies/utils.py", line 801, in get_body_field
    check_file_field(final_field)
  File "/opt/homebrew/lib/python3.11/site-packages/fastapi/dependencies/utils.py", line 100, in check_file_field
    raise RuntimeError(multipart_not_installed_error) from None
RuntimeError: Form data requires "python-multipart" to be installed. 
You can install "python-multipart" with: 

pip install python-multipart


</details>

![image](https://github.com/tiangolo/fastapi/assets/60773912/fa765d3b-0c3c-413e-b7fe-43a9c65da2a1)

I was confused and until finally remembered `python -m` at end

It worked successfully!

I don't know why

<details>
<summary>Test code</summary>

```python
from fastapi import FastAPI, File, UploadFile  # noqa
from fastapi import APIRouter
from fastapi.staticfiles import StaticFiles

app = FastAPI(debug=True)
app.mount('/', StaticFiles(directory='./static'), name='/')

file = APIRouter(prefix='/file', tags=['fileOper'])


@file.post('/upload')
def create_file(file: bytes = File()):
    return dict(code=200, msg='', data=dict(length=len(file)))


@app.get('/')
def index():
    return {"message": "Hello World"}


app.include_router(file)
```
</details>

**Comment by CherrySuryp** [comment]
I encountered same issue, and the problem was caused by conda environment. I've been using both PyCharm and Conda, and that's what caused the promblem.
If you're using PyCharm and encountered same issue you can try deactivate conda environment using `conda deacticate`

**Reply by ser-arthur** [reply]
this really is the issue. worked for me, thanks!

**Comment by rikkkx** [comment]
I have a similar issue. FastAPI can not detect **python-multipart** package.

```bash
❯ python -V
Python 3.12.2

❯ pip freeze | grep fastapi
fastapi==0.115.3

❯ pip freeze | grep multipart
python-multipart==0.0.14
```

As you see, I have **python-multipart** installed on my environment, but still fastapi can't detect it.

That's because **python-multipart** is no more imported via **multipart** package (from version `0.0.14`) as it checks in fastapi ([here](https://github.com/fastapi/fastapi/blob/6ede04f876acc00e2cf688d39e8d6c79dcac0230/fastapi/dependencies/utils.py#L99))

```python
from multipart.multipart import parse_options_header
```

I've got same **ImportError** if I import package this way. From version **0.0.14** package should be imported via **python_multipart**

```python
>>> from multipart.multipart import parse_options_header
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'multipart'
>>> from python_multipart.multipart import parse_options_header
>>> assert parse_options_header
``` 

Can you pls fix this import?

_P.S. as hotfix I can install `python-multipart==0.0.12` and the package will be imported via good old `multipart`_

```bash
❯ pip freeze | grep multipart
python-multipart==0.0.12
❯ python -q
>>> from multipart.multipart import parse_options_header
>>> assert parse_options_header
```

**Reply by enrique-gimenez-itti** [reply]
thanks for this, I faced  the same issue

**Reply by supavit13** [reply]
save my day 👍

**Reply by tuggeluk** [reply]
same here

**Reply by prerakmody** [reply]
Worked for me with `fastapi==0.115.3`

**Reply by pawelad** [reply]
Related issue on the other repo: https://github.com/Kludex/python-multipart/pull/166

**Comment by Chanaka-Prasanna** [comment]
I also faced the same issue. I used Conda env and Pycharm. I did everything mentioned here, but it didn't work. Even when I activated the virtual environment, it pointed out the error to a folder in the C drive. So I installed `python-multipart` globally and used it without virtual environment, and it worked.

But I can not figure out the problem.
